thorlcr/thorutil/thormisc.hpp

```diff
@@ -606,6 +609,118 @@ extern graph_decl bool getExpertOptBool(const char *opt, bool dft=false);
 extern graph_decl __int64 getExpertOptInt64(const char *opt, __int64 dft=0);
 extern graph_decl StringBuffer &getExpertOptString(const char *opt, StringBuffer &out);
 
+////
+// IContextLogger
+class CThorContextLogger : implements IContextLogger, public CSimpleInterface
+{
+    unsigned traceLevel = 1;
+    LogTrace logTrace;
+    mutable CriticalSection statsCrit;
+    mutable CRuntimeStatisticCollection stats;
+
+public:
+    IMPLEMENT_IINTERFACE_USING(CSimpleInterface);
+
+    CThorContextLogger() : stats(jhtreeCacheStatistics, true)
+    {
+        if (globals->hasProp("@httpGlobalIdHeader"))
+            setHttpIdHeaderNames(globals->queryProp("@httpGlobalIdHeader"), globals->queryProp("@httpCallerIdHeader"));
+    }
+    virtual void CTXLOG(const char *format, ...) const override  __attribute__((format(printf,2,3)))
+    {
+        va_list args;
+        va_start(args, format);
+        CTXLOGva(MCdebugProgress, thorJob, NoLogMsgCode, format, args);
+        va_end(args);
+    }
+
+    virtual void CTXLOGva(const LogMsgCategory & cat, const LogMsgJobInfo & job, LogMsgCode code, const char *format, va_list args) const override  __attribute__((format(printf,5,0))) 
+    {
+        VALOG(cat, job, code, format, args);
+    }
+    virtual void logOperatorExceptionVA(IException *E, const char *file, unsigned line, const char *format, va_list args) const __attribute__((format(printf,5,0)))
+    {
+        StringBuffer ss;
+        ss.append("ERROR");
+        if (E)
+            ss.append(": ").append(E->errorCode());
+        if (file)
+            ss.appendf(": %s(%d) ", file, line);
+        if (E)
+            E->errorMessage(ss.append(": "));
+        if (format)
+            ss.append(": ").valist_appendf(format, args);
+        LOG(MCoperatorProgress, thorJob, "%s", ss.str());
+    }
+    virtual void noteStatistic(StatisticKind kind, unsigned __int64 value) const override
+    {
+        DBGLOG("noteStatistic %s = %" I64F "u", queryStatisticName(kind), value);
+        stats.addStatisticAtomic(kind, value);
+    }
+    virtual void setStatistic(StatisticKind kind, unsigned __int64 value) const override
+    {
+        DBGLOG("setStatistic %s = %" I64F "u", queryStatisticName(kind), value);
+        stats.setStatistic(kind, value);
+    }
+    virtual void mergeStats(const CRuntimeStatisticCollection &from) const override
+    {
+        if (from.isThreadSafeMergeSource())
+        {
+            stats.merge(from);
+        }
+        else
+        {
+            CriticalBlock block(statsCrit);
```

Comment:
I think we will need exernal control (outside of CThorContextLogger) of mutex,
because it's the inactiveStats that needs protecting and that needs protecting with the existing activities 'statsCs'